tools/shared.py

```diff
@@ -507,6 +507,16 @@ def get_emscripten_version(path):
 EMSCRIPTEN_VERSION = get_emscripten_version(path_from_root('emscripten-version.txt'))
 parts = [int(x) for x in EMSCRIPTEN_VERSION.split('.')]
 EMSCRIPTEN_VERSION_MAJOR, EMSCRIPTEN_VERSION_MINOR, EMSCRIPTEN_VERSION_TINY = parts
+# For the Emscripten-specific WASM metadata section, follows semver, changes
+# whenever metadata section changes structure
+# NB: major version 0 implies no compatibility
+(EMSCRIPTEN_METADATA_MAJOR, EMSCRIPTEN_METADATA_MINOR) = (0, 0)
+# For the JS/WASM ABI, follows semver, changes whenever musl C types
+# change size/signedness or syscalls change signature. On an ABI
+# change, increment EMSCRIPTEN_ABI_MINOR if EMSCRIPTEN_ABI_MAJOR == 0
+# or the ABI change is backwards compatible, otherwise increment
+# EMSCRIPTEN_ABI_MAJOR and set EMSCRIPTEN_ABI_MINOR = 0
+(EMSCRIPTEN_ABI_MAJOR, EMSCRIPTEN_ABI_MINOR) = (0, 0)
 
```

Comment:
what's the benefit to having both metadata and abi versions?

Reply:
The ABI version is a property of the code inside the wasm file, the metadata version is a property of the structure of the metadata section. Keeping them separate comes down to if whether:

* you want old tools to continue to be able to parse metadata sections when a breaking ABI change happens
* you want to be able to change the metadata structure without changing the ABI version.

I don't see a strong case for the first situation, but I think it might be wise to make it possible to change the metadata structure without bumping the ABI major version (e.g. in the case it's at 1.0).